### ISSUE TITLE
Calculating ratio between fluorescent signals

### ISSUE THREAD

**Issue by GKok** [DillonHammill/CytoExploreR#88]
I recently found out about your package and am really enthusiastic about it - very convenient to use.

Now importing all samples and setting gates works, I'd like to calculate the mean, median and/or modal ratio between two fluorescent signals per sample, eg. divide one by the other, then normalizing the ratio to 1 in a control sample. I can't figure out how to do this - is there an existing function for this? Computing statistics just works on separate signals, not derived parameters/calculations.

Thank you for your fantastic work.

**Comment by DillonHammill** [comment]
@GKok, the current version of CytoExploreR does not have native support for these sorts of operations but this is coming to the next version that will be released soon.

In the meantime, it is easy to calculate statistics such as these manually. See example below:

**Prepare example data**
```{r}
library(CytoExploreR)
library(CytoExploreRData)
library(dplyr)

gs <- GatingSet(Activation)
gs <- cyto_compensate(gs)
gs <- cyto_transform(gs)

gs <- cyto_gatingTemplate_apply(gs, Activation_gatingTemplate)

stats <- cyto_stats_compute(gs,
                            alias = "T Cells",
                            channels = c("CD44", "CD69"),
                            stat = "median",
                            format = "wide")
```

**Compute statistics tidyverse style**
```{r}
# Compute ratio
stats <- stats %>% 
  mutate(ratio = CD44/CD69)
  
# Compute mean ratio for control group
control_ratio <- stats %>%
  filter(Treatment == "Stim-A" & OVAConc == 0) %>%
  summarise(control_ratio = mean(ratio))

# Normalise ratio to control
stats <- stats %>%
  mutate(ratio_norm = ratio/as.numeric(control_ratio))
```

**Comment by GKok** [comment]
Thank you for your quick reply. This would indeed be a temporary solution. However, I do feel this might introduce a slight bias. I'm trying to determine a ratio for each cell, and would like to calculate the median of that ratio only then.

When do you expect the new version to be available? I'll wait for that version for before using it for my final experiments :-)

**Comment by DillonHammill** [comment]
Ah! You are actually after creating derived parameters! You can add derived parameters to the GatingSet as below and then calculate statistics for that parameter:

```{r}
library(CytoExploreR)
library(CytoExploreRData)

gs <- GatingSet(Activation)
cs <- cyto_extract(gs, "root")

lapply(seq_along(cs), function(z){
    new_param <- exprs(cs[[z]])[, "Alexa Fluor 647-A", drop = FALSE]/
        exprs(cs[[z]])[, "7-AAD-A", drop = FALSE]
    colnames(new_param) <- "ratio"
    cf_append_cols(cs[[z]], new_param)
})
```

You will need the latest versions of the RGLab cytoverse packages for this to work.

**Comment by GKok** [comment]
Oh wow, that's it! Now I never need FlowJo again 😄 

Thank you so much for your prompt support!

**Comment by GKok** [comment]
```
> cs <- cyto_extract(gs, "GFP+")
> lapply(seq_along(cs), function(z){
+   new_param <- 
+     exprs(cs[[z]])[, "FL11-A", drop = FALSE] /
+     exprs(cs[[z]])[, "FL1-A", drop = FALSE]
+   colnames(new_param) <- "ratio"
+   cf_append_cols(cs[[z]], new_param)
+ })
 Error in cf_get_h5_file_path(cf) : is(cf, "cytoframe") is not TRUE 
```

I installed the last development version of RGLab cytoverse. I also noticed that you've been in discussion with the author of that package about a similar problem. Do you happen to know if there's a solution at hand?

**Comment by DillonHammill** [comment]
Make sure you are creating the derived parameter for all events (i.e. the "root" node and noy "GFP+").

**Comment by GKok** [comment]
```
> gs
A GatingSet with 61 samples
> cs <- cyto_extract(gs, "root")
> cs
A cytoset with 61 samples.

  column names:
    FSC-H, FSC-A, SSC-H, SSC-A, FL1-H, FL1-A, FL2-H, FL2-A, FL3-H, FL3-A, FL4-H, FL4-A, FL5-H, FL5-A, FL6-H, FL6-A, FL7-H, FL7-A, FL8-H, FL8-A, FL9-H, FL9-A, FL10-H, FL10-A, FL11-H, FL11-A, FL12-H, FL12-A, FL13-H, FL13-A, FSC-Width, Time

> lapply(seq_along(cs), function(z){
+   new_param <- 
+     exprs(cs[[z]])[, "FL11-A", drop = FALSE] /
+     exprs(cs[[z]])[, "FL1-A", drop = FALSE]
+   colnames(new_param) <- "ratio"
+   cf_append_cols(cs[[z]], new_param)
+ })
 Error in cf_get_h5_file_path(cf) : is(cf, "cytoframe") is not TRUE 
```

Unfortunately, that throws the same error...

**Comment by DillonHammill** [comment]
OK I will take a look tomorrow. Did you read in the data from FCS files or are you importing a flowjo workspace?

**Comment by GKok** [comment]
I'm reading FCS files from a Cytoflex S machine. Here's the code I'm using. Reading the samples works fine, drawing and setting gates as well.

Thanks a lot for your help.

```
library("flowCore")
library("flowWorkspace")
library("openCyto")
library("CytoExploreR")
library("cytoverse")
library("cytoqc")

fcs <- load_cytoset_from_fcs(path = path, 
                    #files = allFCS,
                    transformation = FALSE, 
                    alter.names = TRUE, 
                    sep = ";",
                    phenoData = "samples.csv")

gs <- GatingSet(fcs)

# Fluorescent channels
usedChans <- c("FL1-A",  #GFP
               "FL6-A",  #DAPI
               "FL11-A") #RFP

# Transformation
gs <- cyto_transform(gs, channels = usedChans, type="biex")

cyto_gatingTemplate_select("gates.csv")
cyto_gatingTemplate_apply(gs, "gates.csv")
```

**Comment by DillonHammill** [comment]
@GKok, can you supply the output of `sessionInfo()` and `cs[[1]]` so I can troubleshoot this.

**Comment by GKok** [comment]
Definitely!

```
> cs[[1]]
flowFrame object 'SF051 - dFBS 1% Leu - K0.fcs'
with 704941 cells and 32 observables:
            name           desc     range  minRange  maxRange
$P1        FSC-H          FSC-H  16777215         0  16777215
$P2        FSC-A          FSC-A  16777215         0  16777215
$P3        SSC-H          SSC-H  16777215         0  16777215
$P4        SSC-A          SSC-A  16777215      -111  16777215
$P5        FL1-H [488]-525-40-H  16777215         0  16777215
...          ...            ...       ...       ...       ...
$P28      FL12-A [561]-690-50-A  16777215      -111  16777215
$P29      FL13-H [561]-780-60-H  16777215         0  16777215
$P30      FL13-A [561]-780-60-A  16777215      -111  16777215
$P31   FSC-Width      FSC-Width      9999         0      9999
$P32        Time           Time 900000000         0 900000000
627 keywords are stored in the 'description' slot
```

```
> sessionInfo()
R version 4.0.2 (2020-06-22)
Platform: x86_64-apple-darwin17.0 (64-bit)
Running under: macOS  10.16

Matrix products: default
LAPACK: /Library/Frameworks/R.framework/Versions/4.0/Resources/lib/libRlapack.dylib

locale:
[1] nl_NL.UTF-8/nl_NL.UTF-8/nl_NL.UTF-8/C/nl_NL.UTF-8/nl_NL.UTF-8

attached base packages:
[1] stats     graphics  grDevices utils     datasets  methods   base     

other attached packages:
 [1] dplyr_1.0.2               cytoqc_0.99.2             CytoML_2.0.5              ggcyto_1.16.0            
 [5] ncdfFlow_2.34.0           BH_1.72.0-3               RcppArmadillo_0.9.900.3.0 ggplot2_3.3.2            
 [9] cytoverse_0.0.0.9000      CytoExploreR_1.0.8        openCyto_2.0.0            flowWorkspace_4.0.6      
[13] flowCore_2.1.2           

loaded via a namespace (and not attached):
  [1] Rtsne_0.15          colorspace_1.4-1    ellipsis_0.3.1      mclust_5.4.6        cytolib_2.0.3       corpcor_1.6.9      
  [7] base64enc_0.1-3     clue_0.3-57         rstudioapi_0.11     hexbin_1.28.1       IDPmisc_1.1.20      DT_0.15            
 [13] RSpectra_0.16-0     fansi_0.4.1         mvtnorm_1.1-1       xml2_1.3.2          splines_4.0.2       R.methodsS3_1.8.1  
 [19] mnormt_2.0.2        robustbase_0.93-6   knitr_1.30          shinythemes_1.1.2   jsonlite_1.7.1      umap_0.2.6.0       
 [25] cluster_2.1.0       png_0.1-7           R.oo_1.24.0         graph_1.66.0        shiny_1.5.0         httr_1.4.2         
 [31] rrcov_1.5-5         compiler_4.0.2      assertthat_0.2.1    Matrix_1.2-18       fastmap_1.0.1       cli_2.0.2          
 [37] later_1.1.0.1       visNetwork_2.0.9    htmltools_0.5.0     tools_4.0.2         rsvd_1.0.3          gtable_0.3.0       
 [43] glue_1.4.2          reshape2_1.4.4      tinytex_0.26        Rcpp_1.0.5          Biobase_2.48.0      vctrs_0.3.4        
 [49] changepoint_2.2.2   xfun_0.18           stringr_1.4.0       rvest_0.3.6         mime_0.9            lifecycle_0.2.0    
 [55] gtools_3.8.2        XML_3.99-0.5        DEoptimR_1.0-8      zlibbioc_1.34.0     MASS_7.3-53         zoo_1.8-8          
 [61] scales_1.1.1        EmbedSOM_2.1.1      RProtoBufLib_2.1.0  promises_1.1.1      parallel_4.0.2      RBGL_1.64.0        
 [67] RColorBrewer_1.1-2  yaml_2.2.1          gridExtra_2.3       reticulate_1.16     latticeExtra_0.6-29 stringi_1.5.3      
 [73] S4Vectors_0.26.1    pcaPP_1.9-73        flowClust_3.26.0    flowAI_1.18.5       BiocGenerics_0.34.0 flowViz_1.52.0     
 [79] rlang_0.4.7         pkgconfig_2.0.3     matrixStats_0.57.0  evaluate_0.14       fda_5.1.5.1         lattice_0.20-41    
 [85] purrr_0.3.4         ks_1.11.7           htmlwidgets_1.5.1   tidyselect_1.1.0    plyr_1.8.6          magrittr_1.5       
 [91] R6_2.4.1            generics_0.0.2      pillar_1.4.6        withr_2.3.0         tibble_3.0.3        crayon_1.3.4       
 [97] KernSmooth_2.23-17  ellipse_0.4.2       tmvnsim_1.0-2       rhandsontable_0.3.7 rmarkdown_2.4       jpeg_0.1-8.1       
[103] grid_4.0.2          data.table_1.13.0   Rgraphviz_2.32.0    webshot_0.5.2       digest_0.6.25       xtable_1.8-4       
[109] tidyr_1.1.2         httpuv_1.5.4        R.utils_2.10.1      flowStats_4.1.0     openssl_1.4.3       RcppParallel_5.0.2 
[115] stats4_4.0.2        munsell_0.5.0       viridisLite_0.3.0   kableExtra_1.2.1    colortable_0.3.0    askpass_1.1        
```

**Comment by DillonHammill** [comment]
@GKok, apologies for the delayed response. Looks like your `cytolib` is outdated - please update your RGLab packages as below:
```{r}
library(BiocManager)
install(c("cytolib", "flowCore", "flowWorkspace"))
```

Once your package are updated the code should work as expected. If you don't want to update your RGLab packages then the updated code below should work with your current package versions:
```{r}
fcs <- load_cytoset_from_fcs(path = path, 
                    #files = allFCS,
                    transformation = FALSE, 
                    alter.names = TRUE, 
                    sep = ";",
                    phenoData = "samples.csv")

gs <- GatingSet(fcs)

# Fluorescent channels
usedChans <- c("FL1-A",  #GFP
               "FL6-A",  #DAPI
               "FL11-A") #RFP

# Transformation
gs <- cyto_transform(gs, channels = usedChans, type="biex")

cyto_gatingTemplate_select("gates.csv")
cyto_gatingTemplate_apply(gs, "gates.csv")

cs <- cyto_extract(gs, "root")

lapply(seq_along(cs), function(z){
  new_param <- 
    exprs(cs[[z, , returnType = "cytoframe"]])[, "FL11-A", drop = FALSE] /
    exprs(cs[[z, , returnType = "cytoframe"]])[, "FL1-A", drop = FALSE]
  colnames(new_param) <- "ratio"
  cf_append_cols(cs[[z]], new_param)
})
```

**Comment by GKok** [comment]
I installed the latest version of flowCore from Github, but had some trouble installing the latest version of cytolib and therefore flowWorkspaces (https://github.com/RGLab/cytolib/issues/41). Fixed it by updating gfortran to 4.8.x.

After updating all three to the latest Github versions, I still get the same error with the previous code.

```
> cs <- cyto_extract(gs, "root")
> lapply(seq_along(cs), function(z){
+   new_param <- 
+     exprs(cs[[z]])[, "FL11-A", drop = FALSE] /
+     exprs(cs[[z]])[, "FL1-A", drop = FALSE]
+   colnames(new_param) <- "ratio"
+   cf_append_cols(cs[[z]], new_param)
+ })
 Error in cf_get_h5_file_path(cf) : is(cf, "cytoframe") is not TRUE 

> sessionInfo()
R version 4.0.2 (2020-06-22)
Platform: x86_64-apple-darwin17.0 (64-bit)
Running under: macOS  10.16

Matrix products: default
LAPACK: /Library/Frameworks/R.framework/Versions/4.0/Resources/lib/libRlapack.dylib

locale:
[1] nl_NL.UTF-8/nl_NL.UTF-8/nl_NL.UTF-8/C/nl_NL.UTF-8/nl_NL.UTF-8

attached base packages:
[1] stats     graphics  grDevices utils     datasets  methods   base     

other attached packages:
 [1] DataEditR_0.0.6           dplyr_1.0.2               cytoqc_0.99.2             CytoML_2.0.5              ggcyto_1.16.0            
 [6] ncdfFlow_2.34.0           BH_1.72.0-3               RcppArmadillo_0.9.900.3.0 ggplot2_3.3.2             cytoverse_0.0.0.9000     
[11] CytoExploreR_1.0.8        openCyto_2.0.0            flowWorkspace_4.1.10      flowCore_2.1.3           
```

The workaround doesn't return an error. In the output of the lapply() line, it shows all samples with the ratio (range, minRange, maxRange) of all channels, which includes the new 'ratio'. However, when viewing either `cs` or `gs`, the new column/channel doesn't appear to have been linked. I removed some output to shorten it (marked `[...]`).

```
> cs <- cyto_extract(gs, "root")
> cs
A cytoset with 86 samples.

  column names:
    FSC-H, FSC-A, SSC-H, SSC-A, FL1-H, FL1-A, FL2-H, FL2-A, FL3-H, FL3-A, FL4-H, FL4-A, FL5-H, FL5-A, FL6-H, FL6-A, FL7-H, FL7-A, FL8-H, FL8-A, FL9-H, FL9-A, FL10-H, FL10-A, FL11-H, FL11-A, FL12-H, FL12-A, FL13-H, FL13-A, FSC-Width, Time

> lapply(seq_along(cs), function(z){
+   new_param <- 
+     exprs(cs[[z, returnType = "cytoframe"]])[, "FL11-A", drop = FALSE] /
+     exprs(cs[[z, returnType = "cytoframe"]])[, "FL1-A", drop = FALSE]
+   colnames(new_param) <- "ratio"
+   cf_append_cols(cs[[z, returnType = "cytoframe"]], new_param)
+ })

[... samples 1-85 ...]

[[86]]
cytoframe object 'file163d45f2891a5'
with 9126 cells and 33 observables:
          name           desc          range     minRange       maxRange
$P1      FSC-H          FSC-H  16777214.0000    0.0000000  16777214.0000
$P2      FSC-A          FSC-A  16777214.0000    0.0000000  16777214.0000
$P3      SSC-H          SSC-H  16777214.0000    0.0000000  16777214.0000
$P4      SSC-A          SSC-A  16777214.0000    0.0000000  16777214.0000
[...]
$P6      FL1-A [488]-525-40-A      4364.6182  441.6130676      4364.6182
[...]
$P16     FL6-A [405]-450-45-A      4364.6182  441.6130676      4364.6182
[...]
$P26    FL11-A [561]-610-20-A      4364.6182  441.6130676      4364.6182
[...]
$P31 FSC-Width      FSC-Width      9998.0000    0.0000000      9998.0000
$P32      Time           Time 900000000.0000    0.0000000 900000000.0000
$P33     ratio           <NA>         9.6686    0.2622693         9.6686
634 keywords are stored in the 'description' slot

> cs
A cytoset with 86 samples.

  column names:
    FSC-H, FSC-A, SSC-H, SSC-A, FL1-H, FL1-A, FL2-H, FL2-A, FL3-H, FL3-A, FL4-H, FL4-A, FL5-H, FL5-A, FL6-H, FL6-A, FL7-H, FL7-A, FL8-H, FL8-A, FL9-H, FL9-A, FL10-H, FL10-A, FL11-H, FL11-A, FL12-H, FL12-A, FL13-H, FL13-A, FSC-Width, Time

> cyto_markers(gs)
           FSC-H            FSC-A            SSC-H            SSC-A            FL1-H            FL1-A            FL2-H            FL2-A 
         "FSC-H"          "FSC-A"          "SSC-H"          "SSC-A" "[488]-525-40-H" "[488]-525-40-A" "[488]-690-50-H" "[488]-690-50-A" 
           FL3-H            FL3-A            FL4-H            FL4-A            FL5-H            FL5-A            FL6-H            FL6-A 
"[640]-660-20-H" "[640]-660-20-A" "[640]-712-25-H" "[640]-712-25-A" "[640]-780-60-H" "[640]-780-60-A" "[405]-450-45-H" "[405]-450-45-A" 
           FL7-H            FL7-A            FL8-H            FL8-A            FL9-H            FL9-A           FL10-H           FL10-A 
"[405]-525-40-H" "[405]-525-40-A" "[405]-610-20-H" "[405]-610-20-A" "[405]-660-20-H" "[405]-660-20-A" "[561]-585-42-H" "[561]-585-42-A" 
          FL11-H           FL11-A           FL12-H           FL12-A           FL13-H           FL13-A        FSC-Width             Time 
"[561]-610-20-H" "[561]-610-20-A" "[561]-690-50-H" "[561]-690-50-A" "[561]-780-60-H" "[561]-780-60-A"      "FSC-Width"           "Time" 
```

Logically, any attempt towards calculating the median of the new channel, throws an error.

```
> cyto_stats_compute(cs,
+                    alias = "root",
+                    stat = "median",
+                    channels = "ratio",
+                    format = "wide")
Error in FUN(X[[i]], ...) : ratio is not a valid channel/marker.

> cyto_stats_compute(gs,
+                    alias = "root",
+                    stat = "median",
+                    channels = "ratio",
+                    format = "wide")
Error in FUN(X[[i]], ...) : ratio is not a valid channel/marker.

```

Something with `cf_append_cols()`?

**Comment by DillonHammill** [comment]
This should do the trick:
```{r}
fcs <- load_cytoset_from_fcs(path = path, 
                    #files = allFCS,
                    transformation = FALSE, 
                    alter.names = TRUE, 
                    sep = ";",
                    phenoData = "samples.csv")

gs <- GatingSet(fcs)

# Fluorescent channels
usedChans <- c("FL1-A",  #GFP
               "FL6-A",  #DAPI
               "FL11-A") #RFP

# Transformation
gs <- cyto_transform(gs, channels = usedChans, type="biex")

cyto_gatingTemplate_select("gates.csv")
cyto_gatingTemplate_apply(gs, "gates.csv")

cs <- cyto_extract(gs, "root")

# Add derived parameter to each cytoframe
cf_list <- lapply(seq_along(cs), function(z){
  new_param <- 
    exprs(cs[[z, returnType = "cytoframe"]])[, "FL11-A", drop = FALSE] /
    exprs(cs[[z, returnType = "cytoframe"]])[, "FL1-A", drop = FALSE]
  colnames(new_param) <- "ratio"
  cf_append_cols(cs[[z, returnType = "cytoframe"]], new_param)
})
names(cf_list) <- sampleNames(cs)

# Add new cytoframes to a new cytoset
cs_new <- cytoset(cf_list)

# Replace the data stored in GatingSet
gs_cyto_data(gs) <- cs_new
```
FYI the pre-built binaries for RGLab packages should be available from BioConductor, so there is no need to build these packages from source by pulling them from GitHub.

**Comment by GKok** [comment]
Thanks, I've got it to work! 

However, the gatingset seems to lose the phenoData, and all filenames are replaced by 'file[randomnumber]'. However, I can overcome that by replacing these filenames by the 'sampleNames()', as I assume the order stays the same. 

Many thanks for all your support, I really appreciate it. 

PS I pulled the Github versions as the repository versions were those installed and thus outdated.
